hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/TranslateToPublicAddressProvider.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.impl.spi.impl;
+
+import com.hazelcast.client.config.ClientConfig;
+import com.hazelcast.client.impl.connection.AddressProvider;
+import com.hazelcast.client.properties.ClientProperty;
+import com.hazelcast.cluster.Address;
+import com.hazelcast.instance.EndpointQualifier;
+import com.hazelcast.instance.ProtocolType;
+import com.hazelcast.internal.cluster.MemberInfo;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.spi.properties.HazelcastProperties;
+
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+
+class TranslateToPublicAddressProvider {
+    private static final int REACHABLE_ADDRESS_TIMEOUT = 1000;
+    private static final int NON_REACHABLE_ADDRESS_TIMEOUT = 3000;
+    private static final int REACHABLE_CHECK_NUMBER = 3;
+
+    private final ClientConfig config;
+    private final HazelcastProperties properties;
+    private final ILogger logger;
+
+    private volatile boolean translateToPublicAddress;
+
+    TranslateToPublicAddressProvider(ClientConfig config, HazelcastProperties properties, ILogger logger) {
+        this.config = config;
+        this.properties = properties;
+        this.logger = logger;
+    }
+
+    void refresh(AddressProvider addressProvider, Collection<MemberInfo> members) {
+        translateToPublicAddress = resolve(addressProvider, members);
+    }
+
+    private boolean resolve(AddressProvider addressProvider, Collection<MemberInfo> members) {
+        if (!(addressProvider instanceof DefaultAddressProvider)) {
+            return false;
+        }
+
+        String publicIpEnabledProperty = properties.getString(ClientProperty.DISCOVERY_SPI_PUBLIC_IP_ENABLED);
```

Comment:
Makes sense. I will document in the code comment that `null` is intentional. 

Also, I will rearrange the code so that we will be consistent with `HazelcastProperties#getBoolean`